pkg/logutil/zap_text_encoder.go

```diff
@@ -0,0 +1,608 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package logutil
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"math"
+	"strings"
+	"sync"
+	"time"
+	"unicode/utf8"
+
+	"go.uber.org/zap/buffer"
+	"go.uber.org/zap/zapcore"
+)
+
+// DefaultTimeEncoder serializes a time.Time to an human-readable formatted string
+func DefaultTimeEncoder(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
+	s := t.Format("2006/01/02 15:04:05.000 -07:00")
+	if e, ok := enc.(*textEncoder); ok {
+		for _, c := range []byte(s) {
+			e.buf.AppendByte(c)
+		}
+		return
+	}
+	enc.AppendString(s)
+}
+
+// ShortCallerEncoder serializes a caller in file:line format.
+func ShortCallerEncoder(caller zapcore.EntryCaller, enc zapcore.PrimitiveArrayEncoder) {
+	enc.AppendString(getCallerString(caller))
+}
+
+func getCallerString(ec zapcore.EntryCaller) string {
+	if !ec.Defined {
+		return "<unknown>"
+	}
+
+	idx := strings.LastIndexByte(ec.File, '/')
+	buf := _pool.Get()
+	for i := idx + 1; i < len(ec.File); i++ {
+		b := ec.File[i]
+		switch {
+		case b >= 'A' && b <= 'Z':
+			buf.AppendByte(b)
+		case b >= 'a' && b <= 'z':
+			buf.AppendByte(b)
+		case b >= '0' && b <= '9':
+			buf.AppendByte(b)
+		case b == '.' || b == '-' || b == '_':
+			buf.AppendByte(b)
+		default:
+		}
+	}
+	buf.AppendByte(':')
+	buf.AppendInt(int64(ec.Line))
+	caller := buf.String()
+	buf.Free()
+	return caller
+}
+
+// For JSON-escaping; see textEncoder.safeAddString below.
+const _hex = "0123456789abcdef"
+
+var _textPool = sync.Pool{New: func() interface{} {
+	return &textEncoder{}
+}}
+
+var (
+	_pool = buffer.NewPool()
+	// Get retrieves a buffer from the pool, creating one if necessary.
+	Get = _pool.Get
+)
+
+func getTextEncoder() *textEncoder {
+	return _textPool.Get().(*textEncoder)
+}
+
+func putTextEncoder(enc *textEncoder) {
+	if enc.reflectBuf != nil {
+		enc.reflectBuf.Free()
+	}
+	enc.EncoderConfig = nil
+	enc.buf = nil
+	enc.spaced = false
+	enc.openNamespaces = 0
+	enc.reflectBuf = nil
+	enc.reflectEnc = nil
+	_textPool.Put(enc)
+}
+
+type textEncoder struct {
+	*zapcore.EncoderConfig
+	buf            *buffer.Buffer
+	spaced         bool // include spaces after colons and commas
+	openNamespaces int
+
+	// for encoding generic values by reflection
+	reflectBuf *buffer.Buffer
+	reflectEnc *json.Encoder
+}
+
+// NewTextEncoder creates a fast, low-allocation Text encoder. The encoder
+// appropriately escapes all field keys and values.
+func NewTextEncoder(cfg zapcore.EncoderConfig) zapcore.Encoder {
+	return &textEncoder{
+		EncoderConfig: &cfg,
+		buf:           _pool.Get(),
```

Comment:
when to return the buf to the pool?

Reply:
zap will free it after writing encoded the entry. 
https://github.com/uber-go/zap/blob/67bc79d13d155c02fd008f721863ff8cc5f30659/zapcore/core.go#L85-L101